<div dir="ltr"><div><div><div>Hello,<br><br></div>I have used OpenMPI in conjunction with Valgrind for a long time now, and developed a list of suppressions for known false positives over time.<br><br></div>Now I am developing a library for inter-thread communication that is based on using OpenMPI with MPI_THREAD_MULTIPLE support. I have noticed that sending large messages from one thread to another in the same process will cause valgrind to complain about invalid reads. I have narrowed it down to one function being executed on four threads in one process. Attached is a tarball containing the error-reproducing program, valgrind suppression file, and valgrind output.<br>
<br></div><div>The strange thing is that the valgrind error message doesn&#39;t fit the pattern of read-after-free or read-past-the-end. I&#39;d like to know the following:<br>
<br></div><div>1) Should I even worry ? The code doesn&#39;t crash, only valgrind complains. Is it a harmless false positive ?<br></div><div>2) If it is an issue, am I using MPI right?<br></div>
<div>3) If I&#39;m using it right, then what causes this ? some kind of internal buffering issue ?<br><br></div><div>Note that I use Issend, so nothing should be freed until its completely been read (in theory).<br><br></div>

<div>Thank you,<br></div><div><br> </div><div><div><div><div>-- <br><br>Dan Ibanez
</div></div></div></div></div>

